src/main/java/teammates/logic/api/Logic.java

```diff
@@ -785,16 +786,19 @@ public static boolean isCourseArchived(CourseAttributes course, InstructorAttrib
     /**
      * Preconditions: <br>
      * * All parameters are non-null.
+     * <br>
+     * @param alreadyRetrievedData The previously obtained query results to be re-used
      * 
      * @return Details of courses the student is in. CourseData objects
      *         returned contain details of feedback sessions too (except the ones
      *         still AWAITING).
      */
-    public List<CourseDetailsBundle> getCourseDetailsListForStudent(String googleId)
-            throws EntityDoesNotExistException {
+    public List<CourseDetailsBundle> getCourseDetailsListForStudent(String googleId,
+                                              Map<String, StudentAttributes> alreadyRetrievedData)
+                                     throws EntityDoesNotExistException {
         
         Assumption.assertNotNull(ERROR_NULL_PARAMETER, googleId);
-        return coursesLogic.getCourseDetailsListForStudent(googleId);
+        return coursesLogic.getCourseDetailsListForStudent(googleId, alreadyRetrievedData);
```

Comment:
As mentioned before, I don't like this technique because it is a 'too clever' or 'sneaky' workaround. It is not apparent the 2nd parameter is being used as a return value (in C#, there is an `out` keyword that can be used to mark such parameters but there's no such thing in Java) which can result in 'gotchas' later. Perhaps define a new data transfer class to carry both return values as one object?